README
-*- mode:text -*- Time-stamp: "09/11/24 18:03:51 jco"

README for AddressBook
======================

AddresBook is an implementation of an addressbook for Emacs that makes
use of the vCard standard to store contact information.

The homepage of the addressbook is at:

   http://addressbook.nongnu.org

Development resources are hosted in savannah:

   http://savannah.nongnu.org/projects/addressbook


Send bug reports and suggestions to jemarch@gnu.org

End of README.


ChangeLog
2009-12-23  Jose E. Marchesi  <jemarch@gnu.org>

	* abook.el (abook-summary-select-contact): New function.
	(abook-summary-save-cards): New function.

2009-11-29  Jose E. Marchesi  <jemarch@gnu.org>

	* abook.el (abook-sort-cards): Accept an optional parameter FIELD.

2009-11-24  Jose E. Marchesi  <jco@terma.com>

	* vcard.el (vcard-region-decoder-methods): Support the nonstandard
	"b" encoding type.
	(vcard-region-decode-base64): Use the emacs provided
	`base64-decode-region' instead of the custom implementation.

	* abook.el: Display of photograph fixed.

2009-11-19  Jose E. Marchesi  <jemarch@gnu.org>

	* abook.el (abook-contact-display-attribute-regular): Spurious
	blank lines removed.

2009-07-22  Jose E. Marchesi  <jemarch@gnu.org>

	* abook.el (abook-summary-create-contact): Renamed from
	'abook-summary-create-vcard'.
	(abook-summary-mode): Use 'abook-summary-create-contact' and bind
	it to the key 'c'.
	(abook-summary-create-contact): Switch to the contact buffer
	before to call 'abook-summary-create-contact'.

2009-07-13  Jose E. Marchesi  <jemarch@gnu.org>

	* Makefile (SOURCE): Contents are now explicit.

2009-05-06  Jose E. Marchesi  <jemarch@gnu.org>

	* abook.el: Replace abook-create-vcard* with abook-create-contact-*

2009-03-05  Jose E. Marchesi  <jemarch@gnu.org>

	* abook.el (abook-import-vcard): Import all the vcards found in
	the file and not just the first one.
	(abook-import-vcard): Docstring fixed.
	(abook-summary-create-vcard): New function.
	(abook-summary-import-vcard): New function.
	(abook-contact-redisplay-attr-at-point): New function.

	* Makefile: Updated to use the new abook.* files.

	* abook.texi: Generate abook.info instead of addressbook.info
	addressbook -> abook.

	* addressbook.el: Relicensed as GPLv3+.
	Copyright notification updated.
	(addrbook-file): Use ~/.abook as the standard file containing the
	addressbook, since ~/.addressbook is being used by pine.

2007-05-31  Jose E. Marchesi  <jemarch@gnu.org>

	* vcard.el: FSF address updated in copyright notice.

	* COPYING: New file.

2007-05-30  Jose E. Marchesi  <jemarch@gnu.org>

	* uuid.el: New file.
	(uuid): New group.
	(uuid-ifconfig-program): New custom.
	(uuid-hexoctect-regexp): New variable.
	(uuid-time-low-regexp): New variable.
	(uuid-time-mid-regexp): New variable.
	(uuid-version-and-time-high-regexp): New variable.
	(uuid-variant-and-clock-seqhigh-regexp): New variable.
	(uuid-clock-seq-low-regexp): New variable.
	(uuid-node-regexp): New variable.
	(uuid-regexp): New variable.
	(uuid-time-based-version-hex): New variable.
	(uuid-dce-security-version-hex): New variable.
	(uuid-name-based-md5-version-hex): New variable.
	(uuid-name-based-sha1-version-hex): New variable.
	(uuid-random-number-based-version-hex): New variable.
	(uuid-namespace-dns): New variable.
	(uuid-namespace-url): New variable.
	(uuid-namespace-oid): New variable.
	(uuid-namespace-x500): New variable.
	(uuid-generate): New function.
	(uuid-generate-name-based): New function.
	(uuid-generate-random-based): New function.
	(uuid-generate-time-based): New function.
	(uuid-generate-time): New function.
	(uuid-generate-clock-sequence): New function.
	(uuid-format-mac-address): New function.
	(uuid-get-mac-address): New function.
	(uuid-namespace-to-string): New function.
	(uuidp): New function.
	(uuid-lessp): New function.
	(uuid-equal): New function.

2007-05-27  Xavier Maillard  <xma@gnu.org>

	* addressbook.el (addrbook-summary-mode): Bind <up> and <down> to
	respectively previous and next contact. Bind 'a' to contact creation.

2007-05-15  Jose E. Marchesi  <jemarch@gnu.org>

	* addressbook.el (addrbook-backend): New customize.
	(addrbook-directory): New customize.
	(addrbook-be-read-cards): New function.
	(addrbook-be-write-card): New function.
	(addrbook-be-delete-card): New function.

2007-05-10  Xavier Maillard  <xma@gnu.org>

	* addressbook.el (addrbook-contact-display-attribute-photo-logo):
	Introduce url package again so that we can fetch local or remote
	image. Only require url when we are sure we need it and use
	string-as-unibyte to get data.
	(addrbook): Added :link property

2007-05-09  Jose E. Marchesi  <jose.marchesi@deimos-space.com>

	* addressbook.el: outline-mode like headlines

2007-05-08  Xavier Maillard  <xma@gnu.org>

	* vcard.el (vcard-parse-file): New helper function used in addrbook-import-vcard.

	* addressbook.el (addrbook-bury): New function.
	(addrbook-summary-mode, addrbook-contact-mode): Bind b key to
	addrbook-bury.
	(addrbook-write-data-1): New function. Take one VCARD and write
	its data in a FILE.
	(addrbook-import-vcard, addrbook-export-vcard): New functions. Use
	it. Import and Export VCARDs from addrbook-summary buffer.
	(addrbook-summary-mode): Bind i and x to import/export.
	(addrbook-contact-display-attribute-photo-logo): Dropped misuse of
	the url package and replace it with simpler functions. No longer
	require url package.

	* Makefile: Completely rewritten to follow standards. Added
	several new target to build and to install the manual.
	Added license header and CVS tag.

2007-05-08  Jose E. Marchesi  <jose.marchesi@deimos-space.com>

	* addressbook.el (addrbook-summary-display): expand line to
	frame-width. Do not fill with blanks. Append a extra newline to
	the buffer in order to let the line highlight overlay trick to
	work in the last entry.
	(addrbook-summary-goto-contact): Make the contact buffer writable
	before card displaying. Trick to expand the overlay to the entire
	line without filling with blanks.
	(addrbook-summary): Remove call to `addrbook-open'.
	(addrbook-summary-show-contact): Make the contact buffer read only.
	(addrbook-contact-delete-attribute): renamed from
	`addrbook-contact-delete-attribute-type'
	(addrbook-contact-mode): Use `addrbook-contact-delete-attribute'.
	(addressbook): Show summary when the buffer exists.
	(addrbook-multiple-frames-p): New custom variable.
	(addrbook-show-contact): Take care about `addrbook-multiple-frames-p'.

cogre-fa.el
;;; cogre-fa.el --- Finite Automata support for COGRE

;;; Copyright (C) 2009 Jose E. Marchesi

;; Author: Jose E. Marchesi <jemarch@gnu.org>
;; Keywords: automata

;; This file is not part of GNU Emacs.

;; This is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This software is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.
 
;;; Commentary:
;;
;; Provides Finite Automata edition for COGRE.

(require 'cogre)
(require 'cogre-mode)

;;; Code:

(defclass cogre-fa-graph (cogre-graph)
  ()
  "A graph depicting a Finite Automata.")

(defmethod cogre-fa-graph-transition-table ((graph cogre-fa-graph))
  (with-temp-buffer
    (let ((e (oref (this) elements)))
     
      (buffer-substring-no-properties (point-min) (point-max)))))

(defclass cogre-fa-state (cogre-node)
  ((name-default :initform "q")
   (blank-lines-top :initform 0)
   (blank-lines-bottom :initform 0)
   (final-p :initform nil :custom boolean
            :documentation "Boolean indicating if this state is final.")
   (initial-p :initform nil :custom boolean
              :documentation "Boolean indicating if this is the initial state.")
   (alignment :initform center))
  "A Finite Automata state.")

(defmethod cogre-node-slots ((state cogre-fa-state))
  (let ((result nil))
    (if (oref state final-p)
        (list (list (concat "-" (propertize "f" 'face 'cogre-fa-final-face))))
      (if (oref state initial-p)
          (list (list (concat "-" (propertize "i" 'face 'cogre-fa-initial-face))))
        (list (list "--"))))))

(defface cogre-fa-link-label-face '((((class color) (background dark))
                                     (:foreground "blue"))
                                    (((class color) (background light))
                                     (:foreground "blue")))
  "Face used for the link labels."
  :group 'cogre)

(defface cogre-fa-final-face '((((class color) (background dark))
                                (:foreground "red"))
                               (((class color) (background light))
                                (:foreground "red")))
  "Face used for the final attribute display."
  :group 'cogre)

(defface cogre-fa-initial-face '((((class color) (background dark))
                                  (:foreground "yellow"))
                                 (((class color) (background light))
                                  (:foreground "yellow")))
  "Face used for the initial attribute display."
  :group 'cogre)

(define-key cogre-mode-map "l" 'cogre-set-link-label)

(defun cogre-set-link-label (link label)
  "Set the label of the current LINK to LABEL."
  (interactive (let ((l (cogre-link-at-point-interactive)))
                 (list l (read-string "New Label: " ""
                                      nil (oref l label)))))
  (cogre-erase link)
  (if (equal label "")
      (oset link label nil)
    (oset link label label))
  (if (interactive-p)
      (cogre-render-buffer cogre-graph)))

(defclass cogre-labeled-link (cogre-link)
     ((label :initarg :label
             :initform nil
             :type (or null string)
             :custom string
             :documentation "The label of the link.
If non-null, a string containing the label for this link."))
     "A labeled link.")

(defmethod cogre-render ((link cogre-labeled-link))
  "Render LINK in the current graph."
  (call-next-method)
  (with-slots (start end label) link
    (let* ((hd (cogre-node-horizontal-distance start end))
           (vd (cogre-node-vertical-distance start end))
           linkcoords dir)
      ;; Calculate starting points in relation to our attached nodes.
      (if (> (* hd (oref link horizontal-preference-ratio)) vd)
	  ;; In this case, the X delta is larger than the Y delta,
	  ;; so the line is going mostly left/right.
	  (setq linkcoords (cogre-choose-horizontal-link-anchors start end)
		dir 'horizontal)
	(setq linkcoords (cogre-choose-vertical-link-anchors start end)
	      dir 'vertical))
      (let* ((x1 (nth 0 linkcoords))
             (y1 (nth 1 linkcoords))
             (x2 (nth 2 linkcoords))
             (y2 (nth 3 linkcoords))
             (lx (+ x1 (/ (- x2 x1) 2) (% (- x2 x1) 2)))
             (ly (+ y1 (/ (- y2 y1) 2) (% (- y2 y1) 2))))
        (when label
          (if cogre-erase-mode
              (cogre-erase-rectangle lx ly (length label) 1)
            (picture-goto-coordinate lx ly)
            (picture-insert-rectangle
             (list (propertize label 'face 'cogre-fa-link-label-face))
             nil)))))))
            
(defclass cogre-fa-transition (cogre-labeled-link)
  ((symbol :initform "0" :custom string)
   (end-glyph :initform [("/\\ ")
                         ("\\/ ")
                         ("<") (">")])
   (horizontal-preference-ratio :initform 1))
  "This type of link indicates a state transition.")

;;; End of cogre-fa.el


uuid.el
;;; uuid.el --- Universal Unique Identifiers

;; Copyright (C) 2007, 2008, 2009 Jose E. Marchesi

;; Maintainer: Jose E. Marchesi
;; Keywords: standards

;; $Id: uuid.el,v 1.6 2009/11/29 14:15:37 jemarch Exp $

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary

;; This file contain an implementation of the ITU X.667 Recommendation
;; for the generation of Universal Unique Identifiers (also known as
;; Globally Unique Identifiers or GUIDs).
;;
;; Each UUID is a hexadecimal-coded ascii sequence composed by the
;; following fields (separated by the ascii hypen-minus, 45 character,
;; except between the VariantAndClockSeqHigh and ClockSeqLow):
;;
;; - TimeLow                (4 octects => 8 hexadecimal digits)
;; - TimeMid                (2 octects => 4 hexadecimal digits)
;; - VersionAndTimeHigh     (2 octects => 4 hexadecimal digits)
;; - VariantAndClockSeqHigh (1 octect => 2 hexadecimal digits)
;; - ClockSeqLow            (1 octect => 2 hexadecimal digits)
;; - Node                   (6 octects => 12 hexadecimal digits)
;;
;; For example:
;;
;;  00000000-0000-0000-0000-000000000000
;;
;; There are three standarized ways to generate the values of those
;; fields:
;;
;;   - time-based
;;   - random-based
;;   - name-based
;;
;; You can specify the generation semantics to use via the optional
;; `uuid-type' parameter to `uuid-generate'. The default method is the
;; time-based one.
;;
;; Note that, according to the ITU recommendation, uuid generators
;; should produce lower-case letters in hexadecimal encoding. On the
;; other hand, it is recommended for uuid consumers to be
;; case-insensitive regarding alphabetic characters in hex
;; strings. This implementation follows both recommendations.

;;; Code:

(require 'calc)

(defgroup uuid nil
  "Universal Unique Identifiers"
  :group 'development
  :link '(url-link "http://www.emacswiki.org/cgi-bin/wiki/UniversalUniqueIdentifiers"))

(defcustom uuid-ifconfig-program
  "/sbin/ifconfig"
  "Location of the `ifconfig' program to determine the MAC
address to use in the time-based method. If it is set to nil,
then a standarized alternative random method is used."
  :group 'uuid)

(defvar uuid-hexoctect-regexp
  "[0-9a-fA-F][0-9a-fA-F]"
  "Regexp that matches the hexadecimal representation of an octect using lower-case letters")

(defvar uuid-time-low-regexp
  (concat uuid-hexoctect-regexp
          uuid-hexoctect-regexp
          uuid-hexoctect-regexp
          uuid-hexoctect-regexp)
  "Regexp that matches the TimeLow field of a uuid")

(defvar uuid-time-mid-regexp
  (concat uuid-hexoctect-regexp
          uuid-hexoctect-regexp)
  "Regexp that matches the TimeMid field of a uuid")

(defvar uuid-version-and-time-high-regexp
  (concat uuid-hexoctect-regexp
          uuid-hexoctect-regexp)
  "Regexp that matches the VersionAndTimeHigh field of a uuid")

(defvar uuid-variant-and-clock-seqhigh-regexp
  uuid-hexoctect-regexp
  "Regexp that matches the VariantAndClockSeqHigh field of a uuid")

(defvar uuid-clock-seq-low-regexp
  uuid-hexoctect-regexp
  "Regexp that matches the ClockSeqLow field of a uuid")

(defvar uuid-node-regexp
  (concat uuid-hexoctect-regexp
          uuid-hexoctect-regexp
          uuid-hexoctect-regexp
          uuid-hexoctect-regexp
          uuid-hexoctect-regexp
          uuid-hexoctect-regexp)
  "Regexp that matches the Node field of a uuid")

(defvar uuid-regexp
  (concat "^"
          uuid-time-low-regexp
          "-"
          uuid-time-mid-regexp
          "-"
          uuid-version-and-time-high-regexp
          "-"
          uuid-variant-and-clock-seqhigh-regexp
          uuid-clock-seq-low-regexp
          "-"
          uuid-node-regexp
          "$")
  "Regexp that matches a uuid hexadecimal-coded value")

(defvar uuid-time-based-version-hex
  "1"
  "Hexadecimal string encoding the time-based version of a uuid")

(defvar uuid-dce-security-version-hex
  "2"
  "Hexadecimal string encoding the reserved DCE security version of a uuid")

(defvar uuid-name-based-md5-version-hex
  "3"
  "Hexadecimal string encoding the name-based version with MD5 hash of a uuid")

(defvar uuid-name-based-sha1-version-hex
  "4"
  "Hexadecimal string encoding the name-based version with SHA-1 hash of a uuid")

(defvar uuid-random-number-based-version-hex
  "5"
  "Hexadecimal string encoding the random-number-based version of a uuid")

(defvar uuid-namespace-dns
  ;; 6BA7B810
  (list #x6BA7 #xB810
        #x9DAD
        #x11D1
        #x80B4 #x00C0 #x4FD4 #x30C8)
  "ITU X.667 recommended namespace for DNS names")

(defvar uuid-namespace-url
  (list #x6BA7 #xB811
        #x9DAD
        #x11D1
        #x80B4 #x00C0 #x4FD4 #x30C8)
  "ITU X.667 recommended namespace for URL names")

(defvar uuid-namespace-oid
  (list #x6BA7 #xB812
        #x9DAD
        #x11D1
        #x80B4 #x00C0 #x4FD4 #x30C8)
  "ITU X.667 recommended namespace for OID names")

(defvar uuid-namespace-x500
  (list #x6BA7 #xB814
        #x9DAD
        #x11D1
        #x80B4 #x00C0 #x4FD4 #x30C8)        
  "ITU X.667 recommended namespace for directory names")

;;;###autoload
(defun uuid-generate (&optional uuid-type namespace name)
  "Generate and return a new universal unique identifier according
with the ITU X.667 Recommendation for the generation of Universal Unique
Identifiers.

If specified, UUID-TYPE identifies the desired uuid type: `time',
`name-md5', `name-sha1' or `random'. It defaults to `time'.

If specified and `name-md5' or `name-sha1' is used, NAMESPACE is
the namespace to use (see `uuid-namespace-XXX' variables).

If specified, NAME is the name for the `name-md5' or `name-sha1'
method."
  (if (not uuid-type)
      (setq uuid-type 'time))
  (cond
   ((equal uuid-type 'time)
    (uuid-generate-time-based))
   ((or (equal uuid-type 'name-md5)
        (equal uuid-type 'name-sha1))
    (when (not (and namespace name))
      (error "You must specify values for both NAMESPACE and NAME"))
    (uuid-generate-name-based uuid-type namespace name))
   ((equal uuid-type 'random)
    (uuid-generate-random-based))
   (t
    (error "Wrong generation algorithm.\
 Valid ones are 'time 'name-md5 'name-sha1 or 'random"))))

(defun uuid-generate-name-based (type namespace name)
  "Generate and return a name-based uuid."
  (let (time-low
        time-mid
        version-and-time-high
        clock-seq-low
        variant-and-clock-seq-high
        node
        hash
        (name-sequence "")
        i)
    ;; Convert the name to a canonical sequence of octets (as defined by the standards or conventions of its
    ;; name space).
    (dotimes (i (length name))
      (setq name-sequence
            (concat name-sequence (format "%.2x" (aref name i)))))
    ;; Compute the 16-octet hash value of the name space identifier
    ;; concatenated with the name, using the hash function specified
    ;; in 14.2 or 14.3. The numbering of the octets in the hash value
    ;; is from 0 to 15, as specified in IETF RFC 1321 (for MD5) and as
    ;; specified in FIPS PUB 180-2 for SHA-1.
    (cond
     ((equal type 'name-sha1) 
      (error "Name-based type method sha1 not implemented"))
     ((equal type 'name-md5)
      (setq hash (md5 (concat (uuid-namespace-to-string namespace)
                              name-sequence))))
     (t
      (error "Wrong name-based type")))

    ;; Set octets 3 through 0 of the "TimeLow" field to octets 3
    ;; through 0 of the hash value.
    (setq time-low 
          (concat (format "%.2x" 0)
                  (substring hash 26)))
    ;; Set octets 1 and 0 of the "TimeMid" field to octets 5 and 4 of
    ;; the hash value.
    (setq time-mid
          (substring hash 22 25))
    ;; Set octets 1 and 0 of the "VersionAndTimeHigh" field to octets
    ;; 7 and 6 of the hash value.

    ;; Overwrite the four most significant bits (bits 15 through 12)
    ;; of the "VersionAndTimeHigh" field with the four-bit version
    ;; number from Table 3 of 12.2 for the hash function that was
    ;; used.

    ;; Set the "VariantAndClockSeqHigh" field to octet 8 of the hash
    ;; value.

    ;; Overwrite the two most significant bits (bits 7 and 6) of the
    ;; "VariantAndClockSeqHigh" field with 1 and 0, respectively.

    ;; Set the "ClockSeqLow" field to octet 9 of the hash value.

    ;; Set octets 5 through 0 of the "Node" field to octets 15 through
    ;; 10 of the hash value.
    (concat time-low "-"
            time-mid "-"
            version-and-time-high "-"
            variant-and-clock-seq-high
            clock-seq-low "-"
            node)))
  
(defun uuid-generate-random-based ()
  "Generate and return a random-based uuid"
  (let (time-low 
        time-mid
        version-and-time-high
        clock-seq-low
        variant-and-clock-seq-high
        node)
    ;; Set the two most significant bits (bits 7 and 6) of the
    ;; "VariantAndClockSeqHigh" field to 1 and 0, respectively.
    (setq variant-and-clock-seq-high
          (format "%.2x" (logior #x80 (logand #xBF (random (expt 2 8))))))
    ;; Set the four most significant bits (bits 15 through 12) of the
    ;; "VersionAndTimeHigh" field to the four-bit version number
    ;; specified in 12.2.
    (setq version-and-time-high 
          (concat uuid-random-number-based-version-hex
                  (format "%.3x" (random (expt 2 12)))))
    ;; Set all the other bits of the UUID to randomly (or
    ;; pseudo-randomly) generated values.
    (setq time-low 
          (concat (format "%.4x" (random (expt 2 16)))
                  (format "%.4x" (random (expt 2 16)))))
    (setq time-mid 
          (format "%.4x" (random (expt 2 16))))
    (setq clock-seq-low 
          (format "%.2x" (random (expt 2 8))))
    (setq node
          (concat
           (format "%.4x" (random (expt 2 16)))
           (format "%.4x" (random (expt 2 16)))
           (format "%.4x" (random (expt 2 16)))))
    (concat time-low "-"
            time-mid "-"
            version-and-time-high "-"
            variant-and-clock-seq-high
            clock-seq-low "-"
            node)))

(defun uuid-generate-time-based ()
  "Generate and return a time-based uuid"
  ;; Determine the values for the UTC-based Time and the Clock
  ;; Sequence to be used in the UUID, as specified in 12.3 and 12.4.
  (let (time 
        clock-sequence
        time-low time-mid version-and-time-high clock-seq-low
        variant-and-clock-seq-high
        node)
    ;; For the purposes of this algorithm, consider Time to be a
    ;; 60-bit unsigned integer and the Clock Sequence to be a 14-bit
    ;; unsigned integer. 
    (setq time (uuid-generate-time))
    (setq clock-sequence (uuid-generate-clock-sequence))
    ;; Set the "TimeLow" field equal to the least significant 32 bits
    ;; (bits 31 through 0) of Time in the same order of significance.
    (setq time-low 
          (concat 
           (format "%.3x" (logand #x000FF (nth 1 time)))
           (format "%.5x" (nth 2 time))))
    ;; Set the "TimeMid" field equal to bits 47 through 32 from the
    ;; Time in the same order of significance.
    (setq time-mid
          (concat 
           (format "%.2x" (logand #x0007F (nth 0 time)))
           (format "%.2x" (ash (nth 1 time) -12))))
                  
    ;; Set the 12 least significant bits (bits 11 through 0) of the
    ;; "VersionAndTimeHigh" field equal to bits 59 through 48 from
    ;; Time in the same order of significance.
    ;; Set the four most significant bits (bits 15 through 12) of the
    ;; "VersionAndTimeHigh" field to the four-bit version number
    ;; specified in 12.2.
    (setq version-and-time-high 
          (concat uuid-time-based-version-hex
                  (format "%.3x" (ash (nth 0 time) -7))))
    ;; Set the "ClockSeqLow" field to the eight least significant bits
    ;; (bits 7 through 0) of the Clock Sequence in the same order of
    ;; significance.
    (setq clock-seq-low 
          (format "%.2x" (logand #x000F clock-sequence)))
    ;; Set the six least significant bits (bits 5 through 0) of the
    ;; "VariantAndClockSeqHigh" field to the six most significant bits
    ;; (bits 13 through 8) of the Clock Sequence in the same order of
    ;; significance.
    ;; Set the two most significant bits (bits 7 and 6) of the
    ;; "VariantAndClockSeqHigh" clock to one and zero, respectively.
    (setq variant-and-clock-seq-high
          (format "%.2x" (logand #x00BF (ash clock-sequence -9))))
    ;; Set the node field to the 48-bit MAC address in the same order
    ;; of significance as the address.
    (let ((mac-address (uuid-get-mac-address)))
      (if mac-address
          (setq node (uuid-format-mac-address mac-address))
        ;; Use a random number
        (setq node 
              (concat
               (format "%.4x" (random (expt 2 16)))
               (format "%.4x" (random (expt 2 16)))
               (format "%.4x" (random (expt 2 16)))))))
    (concat time-low "-"
            time-mid "-"
            version-and-time-high "-"
            variant-and-clock-seq-high
            clock-seq-low "-"
            node)))
  
(defun uuid-generate-time ()
  "Return the number of 100 nanosecond intervals of UTC since the beginning
of the Gregorian calendar (00:00:00, 15 October 1582).

The returned value is a list:

   (TIME-HIGH TIME-MID TIME-LOW)

with three 20-bits unsigned integers that conform a 60-bit
unsigned integer.

NOTE: we use a resolution of seconds in this code."
  ;; 100 ns intervals offset between Gregorian beginning (00:00:00, 15
  ;; October 1582) and the epoch (00:00:00, 1 January 1970):
  ;; 0x1B21DD213814000

  ;; Operate with 20-bit numbers (GNU Emacs assures integers are 
  ;; at least 29 bits wide and 20/4 = 5)
  (let ((greg-epoch-offset-high #x1B21D)
        (greg-epoch-offset-mid #xD2138)
        (greg-epoch-offset-low #x14000)
        since-epoch-high since-epoch-mid since-epoch-low
        (current-time (current-time)))
    ;; Calculate time since the epoch in seconds
    (setq since-epoch-time-low (+ (nth 1 current-time)
                                  (logand #xF (nth 0 current-time))))
    (setq since-epoch-time-mid (ash (nth 0 current-time) -4))
    (setq since-epoch-time-high (ash (nth 2 current-time) -9))
    ;; TODO: since-epoch-time * 10.000.000
    ;; TODO: Finishme
    (list since-epoch-time-high
          since-epoch-time-mid
          since-epoch-time-low)))

(defun uuid-generate-clock-sequence ()
  "Return a clock sequence number that should be interpreted
as a 14-bit unsigned integer.

NOTE: Since this implementation does not store any state, we
follow the ITU recommendation in using a pseudo-random number
that is _not_ derivated from the Node."
  (random (expt 2 14)))

(defun uuid-format-mac-address (mac-addr)
  "Format MAC-ADDR (a valid MAC address) to a raw hex format"
  (downcase (replace-regexp-in-string ":" "" mac-addr)))
   
(defun uuid-get-mac-address ()
  "Return a suitable MAC address from a network card in the host computer.
If no MAC address is found, then return nil."
  (when (file-executable-p uuid-ifconfig-program)
    (save-excursion
      (with-temp-buffer 
        (call-process uuid-ifconfig-program nil t nil "-a")
        (goto-char (point-min))
        (when (re-search-forward "HWaddr " nil t)
          (re-search-forward (concat uuid-hexoctect-regexp
                                     ":"
                                     uuid-hexoctect-regexp
                                     ":"
                                     uuid-hexoctect-regexp
                                     ":"
                                     uuid-hexoctect-regexp
                                     ":"
                                     uuid-hexoctect-regexp
                                     ":"
                                     uuid-hexoctect-regexp) nil t)
          (buffer-substring (match-beginning 0)
                            (match-end 0)))))))
                                
(defun uuid-namespace-to-string (namespace)
  "Return the hex string representation of NAMESPACE"
  (concat
   (format "%.4x" (nth 0 namespace))
   (format "%.4x" (nth 1 namespace))
   (format "%.4x" (nth 2 namespace))
   (format "%.4x" (nth 3 namespace))
   (format "%.4x" (nth 4 namespace))
   (format "%.4x" (nth 5 namespace))
   (format "%.4x" (nth 6 namespace))
   (format "%.4x" (nth 7 namespace))))
  
;;;###autoload
(defun uuidp (uuid)
  "Return t if UUID is a valid uuid"
  (save-match-data
    (when (string-match uuid-regexp uuid)
      t)))

;;;###autoload
(defun uuid-lessp (uuid1 uuid2)
  "Return t if UUID1 is lesser than UUID2."
  (string-lessp uuid1 uuid2))

;;;###autoload
(defun uuid-equal (uuid1 uuid2)
  "Return t if UUID1 and UUID2 are the same uuid."
  (string-equal uuid1 uuid2))

(provide 'uuid)

;;; uuid.el ends here
